File: content-solutions/binary-search/nearest-station.js
function solution(stations, homes) {
  const s = [...stations].sort((a, b) => a - b);
  const lowerBound = (x) => {
    let lo = 0;
    let hi = s.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (s[mid] >= x) hi = mid;
      else lo = mid + 1;
    }
    return lo;
  };
  return homes.map((h) => {
    const i = lowerBound(h);
    let best = Infinity;
    if (i < s.length) best = s[i] - h;
    if (i > 0) best = Math.min(best, h - s[i - 1]);
    return best;
  });
}

File: content-solutions/binary-search/nearest-station.java
import java.util.*;

class Solution {
    static int lowerBound(int[] a, long x) {
        int lo = 0, hi = a.length;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (a[mid] >= x) hi = mid;
            else lo = mid + 1;
        }
        return lo;
    }

    public int[] solution(int[] stations, int[] homes) {
        int[] s = stations.clone();
        Arrays.sort(s);
        int[] answer = new int[homes.length];
        for (int k = 0; k < homes.length; k++) {
            int h = homes[k];
            int i = lowerBound(s, h);
            int best = Integer.MAX_VALUE;
            if (i < s.length) best = s[i] - h;
            if (i > 0) best = Math.min(best, h - s[i - 1]);
            answer[k] = best;
        }
        return answer;
    }
}
